File: Java/linkedlist/DoublyLinkedList.java
package linkedlist;

public class DoublyLinkedList<E> {
	private class DoublyNode<E> {
		private E data;
		private DoublyNode<E> next;
		private DoublyNode<E> previous;
		
		private DoublyNode(E e) {
			data = e;
			next = null;
			previous = null;
		}
	}
	
	private DoublyNode<E> head;
	private DoublyNode<E> tail;
	private int size;
	
	public DoublyLinkedList() {
		head = null;
		tail = null;
		size = 0;
	}
	
	public DoublyLinkedList(E[] elems) {
		this();
		for (E e : elems) {
			addLast(e);
		}
	}
	
	public int size() {
		return size;
	}
	
	public boolean isEmpty() {
		return size == 0;
	}
	
	public E getFirst() {
		if (head == null) {
			return null;
		}
		return head.data;
	}
	
	public E getLast() {
		if (tail == null) {
			return null;
		}
		return tail.data;
	}
	
	public E get(int index) {
		if (index == 0) {
			return getFirst();
		}
		else if (index == size - 1) {
			return getLast();
		}
		else {
			DoublyNode<E> current = head;
			for (int i = 0; i < index; i++) {
				current = current.next;
			}
			return current.data;
		}
	}
	
	public E setFirst(E e) {
		E data = getFirst()
		if (size == 0) {
			addFirst(e);
		}
		else {
			head.data = e;
		}
		return data;
	}
	
	public void setLast(E e) {
		E data = getLast();
		if (size == 0) {
			addLast(e);
		}
		else {
			tail.data = e;
		}
		return data;
	}
	
	public E set(int index, E e) {
		if (index == 0) {
			return setFirst(e);
		}
		else if (index == size - 1) {
			return setLast(e);
		}
		else {
			DoublyNode<E> current = head;
			for (int i = 0; i < index; i++) {
				current = current.next;
			}
			E data = current.data;
			current.data = e;
			return data;
		}
	}
	
	public void addFirst(E e) {
		if (size == 0) {
			head = new DoublyNode(e);
			tail = head;
		}
		else {
			DoublyNode<E> temp = new DoublyNode(e);
			temp.next = head;
			head.previous = temp;
			head = temp;
		}
		size++;
	}
	
	public void addLast(E e) {
		if (size == 0) {
			tail = new DoublyNode(e);
			head = tail;
		}
		else {
			DoublyNode<E> temp = new DoublyNode(e);
			tail.next = temp;
			temp.previous = tail;
			tail = temp;
		}
		size++;
	}
	
	public void add(int index, E e) {
		if (index == 0) {
			addFirst(e);
		}
		else if (index == size - 1) {
			addLast(e);
		}
		else {
			DoublyNode<E> current = head;
			for (int i = 0; i <= index; i++) {
				current = current.next;
			}
			DoublyNode<E> temp = new DoublyNode(e);
			current.previous.next = temp;
			temp.previous = current.previous;
			temp.next = current;
			size++;
		}
	}
	
	public E removeFirst() {
		E data = getFirst();
		if (head == tail) {
			clear();
		}
		else {
			head = head.next;
			head.previous = null;
			size--;
		}
		return data;
	}
	
	public E removeLast() {
		E data = getLast();
		if (head == tail) {
			clear();
		}
		else {
			tail = tail.previous;
			tail.next = null;
			size--;
		}
		return data;
	}
	
	public E remove(int index) {
		if (index == 0) {
			return removeFirst();
		}
		else if (index == size - 1) {
			return removeLast();
		}
		else {
			DoublyNode<E> current = head;
			for (int i = 0; i <= index; i++) {
				current = current.next;
			}
			current.previous.next = current.next;
			current.next.previous = current.previous;
			size--;
			return current.data;
		}
	}
	
	public boolean contains(E e) {
		DoublyNode<E> current = head;
		while (current != null) {
			if (current.data.equals(e)) {
				return true;
			}
			current = current.next;
		}
		return false;
	}
	
	public int indexOf(E e) {
		DoublyNode<E> current = head;
		for (int i = 0; current != null; i++) {
			if (current.data = e) {
				return i;
			}
			current = current.next;
		}
		return -1;
	}
	
	public boolean erase(E e) {
		DoublyNode<E> current = head;
		while (current != null) {
			if (current.data.equals(e)) {
				if (current == head) {
					removeFirst();
				}
				else if (current == tail) {
					removeLast();
				}
				else {
					current.previous.next = current.next;
					current.next.previous = current.previous;
					size--;
				}
				return true;
			}
			current = current.next;
		}
		return false;
	}
	
	public void clear() {
		head = null;
		tail = null;
		size = 0;
	}
}

File: C/linkedlist/doublylinkedlist.c
#include <stdio.h>
#include <stdlib.h>
#include "doublylinkedlist.h"

doublynode* doublynode_new(void* data) {
	doublynode* node = NULL;
	node = malloc(sizeof(doublynode));
	node->data = data;
	node->next = NULL;
	node->previous = NULL;
	return node;
}

doublylinkedlist* doublylinkedlist_new() {
	doublylinkedlist* list = NULL;
	list = malloc(sizeof(doublylinkedlist));
	list->head = NULL;
	list->tail = NULL;
	list->size = 0;
	return list;
}

void doublylinkedlist_init(doublylinkedlist* list) {
	list->head = NULL;
	list->tail = NULL;
	list->size = 0;
}

void doublylinkedlist_free(doublylinkedlist* list) {
	if (list->size == 0) {
		free(list);
		return;
	}
	doublynode* current = list->head;
	doublynode* temp = current->next;
	while (temp != NULL) {
		free(current);
		current = temp;
		temp = temp->next;
	}
	free(current);
	free(list);
}

int doublylinkedlist_size(const doublylinkedlist* list) {
	return list->size;
}

int doublylinkedlist_empty(const doublylinkedlist* list) {
	return list->size == 0;
}

void* doublylinkedlist_getfirst(const doublylinkedlist* list) {
	if (list->head == NULL) {
		return NULL;
	}
	return list->head->data;
}

void* doublylinkedlist_getlast(const doublylinkedlist* list) {
	if (list->tail == NULL) {
		return NULL;
	}
	return list->head->data;
}

void* doublylinkedlist_get(const doublylinkedlist* list, int index) {
	if (index == 0) {
		return doublylinkedlist_getfirst(list);
	}
	else if (index == list->size - 1) {
		return doublylinkedlist_getlast(list);
	}
	doublynode* current = list->head;
	int i;
	for (i = 0; i < index; ++i) {
		current = current->next;
	}
	return current->data;
}

void* doublylinkedlist_setfirst(doublylinkedlist* list, void* e) {
	void* data = doublylinkedlist_getfirst(list);
	if (list->size == 0) {
		doublylinkedlist_addfirst(list, e);
	}
	else {
		list->head->data = e;
	}
	return data;
}

void* doublylinkedlist_setlast(doublylinkedlist* list, void* e) {
	void* data = doublylinkedlist_getlast(list);
	if (list->size == 0) {
		doublylinkedlist_addlast(list, e);
	}
	else {
		list->tail->data = e;
	}
	return data;
}

void* doublylinkedlist_set(doublylinkedlist* list, int index, void* e) {
	if (index == 0) {
		return doublylinkedlist_setfirst(list, e);
	}
	else if (index == list->size - 1) {
		return doublylinkedlist_setlast(list, e);
	}
	doublynode* current = list->head;
	int i;
	for (i = 0; i < index; ++i) {
		current = current->next;
	}
	void* data = current->data;
	current->data = e;
	return data;
}

void doublylinkedlist_addfirst(doublylinkedlist* list, void* e) {
	doublynode* temp = doublynode_new(e);
	if (list->size == 0) {
		list->head = temp;
		list->tail = temp;
	}
	else {
		temp->next = list->head;
		temp->previous = NULL;
		list->head = temp;
	}
	++(list->size);
}

void doublylinkedlist_addlast(doublylinkedlist* list, void* e) {
	doublynode* temp = doublynode_new(e);
	if (list->size == 0) {
		list->tail = temp;
		list->head = temp;
	}
	else {
		list->tail->next = temp;
		temp->previous = list->tail;
		list->tail = temp;
	}
	++(list->size);
}

void doublylinkedlist_add(doublylinkedlist* list, int index, void* e) {
	if (index == 0) {
		doublylinkedlist_addfirst(list, e);
		return;
	}
	else if (index == list->size - 1) {
		doublylinkedlist_addlast(list, e);
		return;
	}
	doublynode* current = list->head;
	int i;
	for (i = 0; i <= index; ++i) {
		current = current->next;
	}
	doublynode* temp = doublynode_new(e);
	current->previous->next = temp;
	temp->previous = current->previous;
	temp->next = current;
	++(list->size);
}

void* doublylinkedlist_removefirst(doublylinkedlist* list) {
	void* data = doublylinkedlist_getfirst(list);
	doublynode* current;
	if (list->head == list->tail) {
		doublylinkedlist_clear(list);
	}
	else {
		current = list->head->next;
		free(list->head);
		list->head = current;
		list->head->previous = NULL;
		--(list->size);
	}
	return data;
}

void* doublylinkedlist_removelast(doublylinkedlist* list) {
	void* data = doublylinkedlist_getlast(list);
	doublynode* current;
	if (list->head == list->tail) {
		doublylinkedlist_clear(list);
	}
	else {
		current = list->tail->previous;
		free(list->tail);
		list->tail = current;
		list->tail->next = NULL;
		--(list->size);
	}
	return data;
}

void* doublylinkedlist_remove(doublylinkedlist* list, int index) {
	if (index == 0) {
		return doublylinkedlist_removefirst(list);
	}
	else if (index == list->size - 1) {
		return doublylinkedlist_removelast(list);
	}
	doublynode* current = list->head;
	int i;
	for (i = 0; i < index; ++i) {
		current = current->next;
	}
	current->previous->next = current->next;
	void* data = current->data;
	free(current);
	--(list->size);
	return data;
}

int doublylinkedlist_contains(const doublylinkedlist* list, void* e) {
	doublynode* current = list->head;
	while (current != NULL) {
		if (current->data == e) {
			return 1;
		}
		current = current->next;
	}
	return 0;
}

int doublylinkedlist_index(const doublylinkedlist* list, void* e) {
	doublynode* current = list->head;
	int i;
	for (i = 0; current != NULL; ++i) {
		if (current->data == e) {
			return i;
		}
	}
	return -1;
}

int doublylinkedlist_erase(doublylinkedlist* list, void* e) {
	int index = doublylinkedlist_index(list, e);
	if (index >= 0) {
		doublylinkedlist_remove(list, index);
		return 1;
	}
	return 0;
}

void doublylinkedlist_clear(doublylinkedlist* list) {
	if (list->size == 0) {
		return;
	}
	
	doublynode* current = list->head;
	doublynode* temp = current->next;
	while (temp != NULL) {
		free(current);
		current = temp;
		temp = temp->next;
	}
	free(current);
	
	list->head = NULL;
	list->tail = NULL;
	list->size = 0;
}
